src/csharp/Microsoft.Spark.E2ETest/UdfTests/UdfComplexTypesTests.cs

```diff
@@ -34,24 +35,233 @@ public UdfComplexTypesTests(SparkFixture fixture)
+                    new StructField("arrIds", new ArrayType(new ArrayType(new IntegerType()))),
+                    new StructField(
+                        "arrArrIds",
+                        new ArrayType(new ArrayType(new ArrayType(new IntegerType()))))
+                });
+
+                var data = new GenericRow[]
+                {
+                    new GenericRow(new object[]
+                    {
+                        "Name1",
+                        new int[] { 1, 2, 3 },
+                        new int[][]
+                        {
+                            new int[] { 10, 11 },
+                            new int[] { 12 }
+                        },
+                        new int[][][]
+                        {
+                            new int[][]
+                            {
+                                new int[] { 100, 101 },
+                                new int[] { 102 }
+                            },
+                            new int[][]
+                            {
+                                new int[] { 103 }
+                            }
+                        }
+                    }),
+                    new GenericRow(new object[]
+                    {
+                        "Name2",
+                        null,
+                        null,
+                        null
+                    }),
+                    new GenericRow(new object[]
+                    {
+                        "Name3",
+                        new int[] { 4 },
+                        new int[][]
+                        {
+                            new int[] { 13 },
+                        },
+                        new int[][][]
+                        {
+                            new int[][]
+                            {
+                                new int[] { 104 }
+                            }
+                        }
+                    }),
+                };
+
+                Func<Column, Column, Column, Column, Column> udf =
+                    Udf<string, int[], int[][], int[][][], string>(
+                        (name, ids, arrIds, arrArrIds) =>
+                        {
+                            var sb = new StringBuilder();
+                            sb.Append(name);
+
+                            if (ids != null)
+                            {
+                                sb.Append("|" + string.Join(",", ids));
+                            }
+
+                            if (arrIds != null)
+                            {
+                                sb.Append("|" + string.Join(",", arrIds.SelectMany(i => i)));
+                            }
+
+                            if (arrArrIds != null)
+                            {
+                                sb.Append("|" + string.Join(
+                                    ",", arrArrIds.SelectMany(i => i.SelectMany(j => j))));
+                            }

```

Comment:
Can we simplify the logic here to abstract the `Append` to a function that takes in array and returns the string?